### PR TITLE
Adds group feature for injecting kwargs

### PR DESCRIPTION
**Pull request by elijahbenizzy**
We were doing this in the replacement function. Looks like everythin ghas to go in the replacement function, which makes things slightly trickier, but at least we can centralize the logic in that function.

[Short description explaining the high-level reason for the pull request]

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/function_modifiers/expanders.py [hunk 1]
```diff
@@ -298,9 +308,16 @@ def __missing__(self, key):
...
             return self.specified_docstrings[output_name]
+        doc = fn.__doc__
         if doc is None:
             return None
-        parametrization = self.parametrization[output_name]
+        parameterizations = self.parameterization.copy()
+        if self.PLACEHOLDER_PARAM_NAME in parameterizations:
+            import pdb
+
+            pdb.set_trace()
```
Comment (skrawcz) [inline hamilton/function_modifiers/expanders.py:318]: classic

**Comment by skrawcz** [comment]
Part of me wants the grouped dependency to be dictionary (or list of tuples) and not a straight list. Why? So that someone could do this and get the correct behavior:

```python
@inject(group(source("foo"), source("bar"), source("baz"))
def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
      """One would likely expect the series to be named and thus the output to be "foo", "bar", "baz"... but it's not guaranteed.""""
     return pd.concat(series_list)

@inject(group(source("foo"), source("bar"), source("baz"))
def my_df(series_list: List[Tuple[str, pd.Series]]) -> pd.DataFrame:
      """This would be guaranteed to work""""
     return pd.DataFrame(dict(series_list))
```

**Comment by skrawcz** [comment]
Otherwise @elijahbenizzy should someone be able to do this:

```python
@inject(group(source("foo"), source("foo"), source("foo")))
def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
    """Should we have three? or just one?"""
    return sum(series_list)

constant = pd.Series([1,2,3,4])
@inject(group(source("foo"), source("foo"), source("foo"), value(constant)))
def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
    """Are values allowed?"""
    return sum(series_list)
```

**Comment by elijahbenizzy** [comment]
> Otherwise @elijahbenizzy should someone be able to do this:
> 
> ```python
> @inject(group(source("foo"), source("foo"), source("foo")))
> def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
>     """Should we have three? or just one?"""
>     return sum(series_list)

Ughh, yeah, currently this is broken but not for great reasons. Need to fix + add tests. Was thinking of releasing then fixing -- as this is a weird case. IMO the strangest case that will currently break is:

> @inject(group(source("foo"), source("boo"), source("baz")))
> def my_df(series_list: List[pd.Series], foo: pd.Series) -> pd.DataFrame:
>     """Are values allowed?"""
>     return sum(series_list)

This is busted due to the popping of kwargs when we parse

> constant = pd.Series([1,2,3,4])
> @inject(group(source("foo"), source("boo"), source("baz"), value(constant)))
> def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
>     """Are values allowed?"""
>     return sum(series_list)
> ```

This is fine by me -- it'll just work I think?

**Comment by skrawcz** [comment]
> > Otherwise @elijahbenizzy should someone be able to do this:
> > ```python
> > @inject(group(source("foo"), source("foo"), source("foo")))
> > def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
> >     """Should we have three? or just one?"""
> >     return sum(series_list)
> > ```
> 
> Ughh, yeah, currently this is broken but not for great reasons. Need to fix + add tests. Was thinking of releasing then fixing -- as this is a weird case. 

I would argue against allowing duplicates, since you can achieve that another way more clearly.

> IMO the strangest case that will currently break is:
> 
> > @Inject(group(source("foo"), source("boo"), source("baz")))
> > def my_df(series_list: List[pd.Series], foo: pd.Series) -> pd.DataFrame:
> > """Are values allowed?"""
> > return sum(series_list)
> 
> This is busted due to the popping of kwargs when we parse
> 
> > constant = pd.Series([1,2,3,4])
> > @Inject(group(source("foo"), source("boo"), source("baz"), value(constant)))
> > def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
> > """Are values allowed?"""
> > return sum(series_list)
> 
> This is fine by me -- it'll just work I think?

I don't think we want that to work TBH. What's the "name" of the constant?

**Comment by elijahbenizzy** [comment]
> Part of me wants the grouped dependency to be a dictionary (or list of tuples) and not a straight list. Why? So that someone could do this and get the correct behavior:
> 
> ```python
> @inject(group(source("foo"), source("bar"), source("baz")))
> def my_df(series_list: List[pd.Series]) -> pd.DataFrame:
>     """One would likely expect the series to be named and thus the output to be "foo", "bar", "baz"... but it's not guaranteed."""
>     return pd.concat(series_list)
> 
> @inject(group(source("foo"), source("bar"), source("baz")))
> def my_df(series_list: List[Tuple[str, pd.Series]]) -> pd.DataFrame:
>     """This would be guaranteed to work"""
>     return pd.DataFrame(dict(series_list))
> 
> @inject(group(source("foo"), source("bar"), source("baz")))
> def my_df(series_dict: Dict[str, pd.Series]) -> pd.DataFrame:
>     """This would be guaranteed to work"""
>     return pd.DataFrame(series_dict)
> ```
> 
> Or maybe we can do all the above and know what to do based on the parameter type annotation?

Yeah, so I think we should allow it to be a dict depending on the annotation -- it should be pretty easy. That said, we could just as well do a `grouped_dict` as well as `grouped_list`, so you could have to annotate it in the right way...

**Comment by elijahbenizzy** [comment]
Merging this to consolidate PRs

**Pull request merged**